lib/ansible/modules/cloud/vmware/vmware_guest.py

```diff
@@ -1207,12 +1207,60 @@ def configure_disks(self, vm_obj):
                 elif disk_type == 'eagerzeroedthick':
                     diskspec.device.backing.eagerlyScrub = True
 
-            # which datastore?
+            # Select actual datastore and datastore name while creating data disk for
+            # virtual machine.
+            if vm_obj is None:
+                vm_name = self.params['name'].replace('.', '_')
+            else:
+                vm_name = vm_obj.name
+
             if expected_disk_spec.get('datastore'):
-                # TODO: This is already handled by the relocation spec,
-                # but it needs to eventually be handled for all the
-                # other disks defined
-                pass
+                # User has specified datastore name
+                datastore_name = expected_disk_spec['datastore']
+                datastore = self.cache.find_obj(self.content, [vim.Datastore], datastore_name)
+
+                if datastore is None:
+                    self.module.fail_json(msg="Unable to find datastore named %s specified "
+                                              "in disk configuration at index %s" % (datastore_name,
+                                                                                     disk_index))
+                parent_dc = self.cache.get_parent_datacenter(datastore)
+                vmdk_file_name = self.create_directory_for_datadisk(datacenter=parent_dc,
+                                                                    datastore=datastore,
+                                                                    vm_name=vm_name,
+                                                                    diskspec=diskspec)
+                diskspec.device.backing.datastore = datastore
+                diskspec.device.backing.fileName = vmdk_file_name
+
+            elif 'autoselect_datastore' in expected_disk_spec and expected_disk_spec['autoselect_datastore']:
+                # User has specified autoselect datastore option
+                datastores = self.cache.get_all_objs(self.content, [vim.Datastore])
+                datastores = [x for x in datastores if
+                              self.cache.get_parent_datacenter(x).name == self.params['datacenter']]
+
+                if datastores is None or len(datastores) == 0:
+                    self.module.fail_json(msg="Unable to find a datastore for disk at "
+                                              "index %s when autoselecting" % disk_index)
+                datastore_freespace = 0
+                for ds in datastores:
+                    if (ds.summary.freeSpace > datastore_freespace) or (
+                            ds.summary.freeSpace == datastore_freespace and not datastore):
+                        # If datastore field is provided, filter destination datastores
+                        if 'datastore' in expected_disk_spec \
```

Comment:
Here, 'datastore' can't be in expected_disk_spec as:
```
if expected_disk_spec.get('datastore'):
    …
elif expected_disk_spec.get('autoselect_datastore', False):
   we are here
```

Reply:
ok.